File: 06_breadth-first_search/swift/01_breadth-first_search.swift
import Foundation

public struct Deque<T> {
	private var array = [T]()
	
	public var isEmpty: Bool {
		return array.isEmpty
	}
	
	public var count: Int {
		return array.count
	}
	
	public mutating func enqueue(_ element: T) {
		array.append(element)
	}
	
	public mutating func enqueueFront(_ element: T) {
		array.insert(element, at: 0)
	}
	
	public mutating func dequeue() -> T? {
		if isEmpty {
			return nil
		} else {
			return array.removeFirst()
		}
	}
	
	public mutating func dequeueBack() -> T? {
		if isEmpty {
			return nil
		} else {
			return array.removeLast()
		}
	}
	
	public func peekFront() -> T? {
		return array.first
	}
	
	public func peekBack() -> T? {
		return array.last
	}
}

func persionIsSeller(name: String) -> Bool {
	return name.characters.last == "m"
}

var graph = [String : [String]]()
graph["you"] = ["alice", "bob", "claire"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "jonny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []

func search(name: String) -> Bool {
	var searchQueue = Deque<String>()
	for string in graph[name]! {
		searchQueue.enqueue(string)
	}

	var searched = [String]()
	while !searchQueue.isEmpty {
		let person = searchQueue.dequeue()
		if !searched.contains(person!) {
			if persionIsSeller(name: person!) {
				print("\(person!) is a mango seller!")
				return true
			} else {
				for string in graph[person!]! {
					searchQueue.enqueue(string)
				}
				searched.append(person!)
			}
		}
	}
	
	return false
}

result = search(name: "you")

if !result {
	print("Mango seller Not Found!")
}

File: 06_breadth-first_search/c++/01_breadth-first_search.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

bool is_seller(string& name) {
	return name.back() == 'm';
}


bool search(string name, unordered_map<string, vector<string>>& graph) {
	queue<string> search_queue;
	unordered_set<string> searched;

	// add all friends to search queue
	for (auto&& friend_name : graph.find(name) -> second)
		search_queue.push(friend_name);

	while (!search_queue.empty()) {
		string person = search_queue.front();
		search_queue.pop();

		// only search this person if you haven't already searched them.
		if (searched.find(person) == searched.end()) {
			if (is_seller(person)) {
				cout << person << " is a mango seller!" << endl;
				return true;
			}
			
			// add all friends of a person to search queue
			for (auto&& friend_name : graph.find(person) -> second)
				search_queue.push(friend_name);

			// mark this person as searched
			searched.insert(person);
		}
	}

	return false;
}

int main() {
	std::unordered_map<std::string, std::vector<std::string>> graph;
	graph.insert({"you", {"alice", "bob", "claire"}});
	graph.insert({"bob", {"anuj", "peggy"}});
	graph.insert({"alice", {"peggy"}});
	graph.insert({"claire", {"thom", "jonny"}});
	graph.insert({"anuj", {}});
	graph.insert({"peggy", {}});
	graph.insert({"thom", {}});
	graph.insert({"jonny", {}});

	string name = "you";
	bool result = search(name, graph);
	cout << "Found mango seller: " << result << endl;
}